### PR TITLE
Task execution uses Tooling API directly

### PR DESCRIPTION
**Pull request by donat**

### CODE REVIEW

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java [hunk 1]
```diff
@@ -64,34 +77,52 @@ protected final void runToolingApiJob(IProgressMonitor monitor) {
...
-        Event event = new DefaultExecuteLaunchRequestEvent(processDescription, request);
-        CorePlugin.listenerRegistry().dispatch(event);
-
-        // launch the build
-        request.executeAndWait();
+        final FixedRequestAttributes attributes = getConfigurationAttributes().toFixedRequestAttributes();
+        ProjectConnection connection = null;
+        try {
+            // apply FixedRequestAttributes on connector and establish TAPI connection
+            GradleConnector connector = GradleConnector.newConnector().forProjectDirectory(attributes.getProjectDir());
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: Ideally, Fixed/Transient attributes should have methods to apply themselves to `GradleConnector`/`ProjectConnection`/`ConfigurableLauncher`, so we don't need to repeat ourselves. We can do that in another PR though.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java [hunk 2]
```diff
@@ -64,34 +77,52 @@ protected final void runToolingApiJob(IProgressMonitor monitor) {
...
+
+            // print the applied run configuration settings at the beginning of the console output
+            OutputStreamWriter writer = new OutputStreamWriter(processStreams.getConfiguration());
+            writeFixedRequestAttributes(attributes, writer, monitor);
+
+            // execute the build
+            launcher.run();
+        } finally {
+            if (connection != null) {
+                connection.close();
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: Our `GradleBuild` interface should provide single-use launchers like this:

```
build = CorePlugin.workspaceManager.gradleBuild(FixedAttributes).newBuild(TransientAttributes)
// do more configuration on the build launcher
build.run()
// connection is closed transparently, calling `run` again throws an IllegalStateException
```

Then we can remove most of the boilerplate above, including the try-catch.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java [hunk 3]
```diff
@@ -41,21 +43,32 @@
...
+ * Base class to execute Gradle builds in a job.
  */
 public abstract class BaseLaunchRequestJob extends ToolingApiJob {
 
+    /**
+     * Common interface to retrieve and execute build and test launch requests. This interface is
+     * needed since there's no common {@code run()} method defined on the
+     * {@link LongRunningOperation} subclasses.
+     */
+    public static interface Launcher {
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: I would just go with an abstract `launch` method on `BaseLaunchRequestJob` and add a generic type parameter for the kind of thing it launches. For instance

```
TestLaunchRequestJob extends BaseLaunchRequestJob<TestLauncher> {
  @Override
  TestLauncher createLauncher(GradleBuild, TransientRequestAttributes) {
    // see also my comments below
    gradleBuild.createTestLauncher(transientRequestAttributes)
  }

  @Override
  void launch(TestLauncher) {
    testLauncher.run()
  }
}
```

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java (hunk 2 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: I like the idea, although there's one issue I'd like to discuss. The launcher is created from `ProjectConnection` so for instance the `RunGradleJvmTestLaunchRequestJob` would have this:
```
    @Override
    protected GradleLauncher createLauncher(GradleBuild gradleBuild, TransientRequestAttributes transientAttributes) {
        return gradleBuild.newTestLaunch(transientAttributes, new LaunchConfig<TestLauncher>() {

            @Override
            public void apply(TestLauncher launch) {
                launch.withTests(RunGradleTestLaunchRequestJob.this.testDescriptors);
            }
        });
    }
```

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java (hunk 3 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: I've adjusted the implementation but I think there's still room for improvement.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java (hunk 1 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: Here's the issue for that: https://github.com/eclipse/buildship/issues/410

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java (hunk 2 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: Fixed.

Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: Using a callback is a bit limiting. I see you had to copy-paste some code as a result instead of having it in the base class like before. It also forced you to create a new class (`GradleInvocation`), which is against our goal for removing indirection layers over the TAPI types.

Instead what I was thinking was returning a `BuildLauncher` or `TestLauncher` which automatically closes the underlying `ProjectConnection` after it has first been used. This can for instance be done with a listener that listens for build completion.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/GradleBuild.java [hunk 4]
```diff
@@ -84,4 +88,36 @@
...
     FixedRequestAttributes getRequestAttributes();
+
+    /**
+     * Creates a new Gradle build invocation.
+     *
+     * @param transientAttributes the transient attributes for the invocation
+     * @param config the configuration for the build launcher
+     * @return the object that can run the Gradle build
+     */
+    GradleInvocation newBuildInvocation(TransientRequestAttributes transientAttributes, BuildLauncherConfig config);
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/GradleBuild.java]: We should not introduce new types here. Instead we should just return a `TestLauncher` (the interface from the TAPI).

The underlying project connection can be closed transparently with a progress listener that the `DefaultGradleBuild` registers. When the "build finished" event is received, we can call `ProjectConnection.close()`.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java (hunk 2 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: Unfortunately, there's no way to add event listeners to the launchers to detect when the build is finished. We could solve this by creating delegate classes for the launchers.

Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: Yes, a decorator pattern would work well there.

Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/launch/BaseLaunchRequestJob.java]: Fixed, although I'm not happy with the boilerplate code. Maybe we should just switch to Kotlin ;)

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    ConnectionAwareBuildLauncherDelegate(ProjectConnection connection, BuildLauncher delegate) {
+        super(delegate);
+        this.connection = connection;
+    }
+
+    @Override
+    public void run() {
+        this.delegate.run();
+        this.connection.close();
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java]: needs to be wrapped in try-finally

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    @Override
+    public void run(final ResultHandler<? super Void> resultHandler) {
+        this.delegate.run(new ResultHandler<Void>() {
+
+            @Override
+            public void onComplete(Void result) {
+                resultHandler.onComplete(result);
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java]: try-finally

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                ConnectionAwareBuildLauncherDelegate.this.connection.close();
+            }
+
+            @Override
+            public void onFailure(GradleConnectionException exception) {
+                resultHandler.onFailure(exception);
+            ConnectionAwareBuildLauncherDelegate.this.connection.close();
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java]: try-finally

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLongRunningOperationDelegate.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+abstract class ConnectionAwareLongRunningOperationDelegate<T extends LongRunningOperation> implements LongRunningOperation {
+
+    protected final T delegate;
+
+    ConnectionAwareLongRunningOperationDelegate(T delegate) {
+        this.delegate = delegate;
+    }
+
+    protected static ProjectConnection openConnection(FixedRequestAttributes fixedAttributes) {
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLongRunningOperationDelegate.java]: The connection opening should be done by the `GradleBuild` instance.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+import org.gradle.tooling.ProjectConnection;
+import org.gradle.tooling.ResultHandler;
+import org.gradle.tooling.events.OperationType;
+import org.gradle.tooling.model.Launchable;
+import org.gradle.tooling.model.Task;
+
+import com.gradleware.tooling.toolingmodel.repository.FixedRequestAttributes;
+import com.gradleware.tooling.toolingmodel.repository.TransientRequestAttributes;
+
+final class ConnectionAwareBuildLauncherDelegate extends ConnectionAwareLongRunningOperationDelegate<BuildLauncher> implements BuildLauncher {
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java]: I think you could save a lot of code here by using `java.lang.Proxy` instead of implementing the delegation pattern by hand. The reflection overhead would be negligible, since we don't call these methods in a hot loop.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLongRunningOperationDelegate.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        GradleConnector connector = GradleConnector.newConnector().forProjectDirectory(fixedAttributes.getProjectDir());
+        GradleDistributionWrapper.from(fixedAttributes.getGradleDistribution()).apply(connector);
+        connector.useGradleUserHomeDir(fixedAttributes.getGradleUserHome());
+        return connector.connect();
+    }
+
+    protected void applyRequestAttributes(FixedRequestAttributes fixedAttributes, TransientRequestAttributes transientAttributes) {
```
Comment (oehme) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLongRunningOperationDelegate.java]: This should be done in `GradleBuild`. These decorators should only handling the automatic closing, nothing else.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java (hunk 9 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java]: Implemented with proxy.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java (hunk 5 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java (hunk 6 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java (hunk 7 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareBuildLauncherDelegate.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLongRunningOperationDelegate.java (hunk 8 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLongRunningOperationDelegate.java]: Fixed

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLongRunningOperationDelegate.java (hunk 10 above)
Comment (donat) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/ConnectionAwareLongRunningOperationDelegate.java]: Fixed

**Comment by oehme** [comment]
Excellent cleanup :)

**Comment by donat** [comment]
We are getting there!
